charms/jupyter-ui/src/charm.py

```diff
@@ -25,139 +39,190 @@ def __init__(self, msg, status_type=None):
-                                ],
-                            },
-                        ],
-                    }
-                ],
             },
+        }
+
+        return Layer(layer_config)
+
+    def _update_layer(self) -> None:
+        """Update the Pebble configuration layer (if changed)."""
+        current_layer = self.container.get_plan()
+        new_layer = self._jupyter_ui_layer
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info("Pebble plan updated with new configuration, replaning")
+                self.container.replan()
+            except ChangeError:
+                raise ErrorWithStatus("Failed to replan", BlockedStatus)
+
+    def _upload_files_to_container(self):
+        """Upload required files to container."""
+        self.container.push(
+            "/etc/config/spawner_ui_config.yaml",
+            "spawner_ui_config.yaml",
+            make_dirs=True,
         )
-        self.model.unit.status = ActiveStatus()
+        for file_name, file_content in yaml.safe_load(
+            Path("src/logos-configmap.yaml").read_text()
+        )["data"].items():
+            logo_file = "/src/apps/default/static/assets/logos/" + file_name
+            self.container.push(
+                logo_file,
+                file_content,
+                make_dirs=True,
+            )
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploys K8S resources."""
+        try:
+            self.unit.status = MaintenanceStatus("Creating K8S resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("K8S resources creation failed", BlockedStatus)
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _on_install(self, _):
+        """Perform installation only actions."""
+        try:
+            self._deploy_k8s_resources()
+        except CheckFailed as err:
+            self.model.unit.status = err.status
+        return
+
+    def _on_pebble_ready(self, _):
+        """Configure started container."""
+        if not self._is_container_ready():
+            return
+
+        # upload files to container
+        self._upload_files_to_container()
+
+        # proceed with other actions
+        self.main(_)
+
+    def _on_remove(self, _):
+        """Remove all resources."""
+        self.unit.status = MaintenanceStatus("Removing K8S resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, k8s_resources_manifests)
+        except ApiError as e:
+            self.logger.warning(f"Failed to delete K8S resources, with error: {e}")
+            raise e
```

Comment:
This came up in @NohaIhab's recent PR too.  If we're trying to remove the charm, do we want to fail on errors here or just log them?  If we `raise e` here, the charm will be stopped and Juju wont remove any of the other resources.  I think someone then has to do a `juju remove-application --force`.  That sort of interruption feels harsh for removing resources

On the other hand though, only logging errors here is basically like shouting into a pillow.  Nobody is going to notice those log messages if we immediately remove the charm afterwards - it'll look like things went well